addons/mail/models/mail_activity.py

```diff
@@ -212,33 +227,63 @@ def _onchange_recommended_activity_type_id(self):
 
+        valid = self._filter_access(operation)
+        if valid != self:
+            error_message = (
+                _('The requested operation cannot be completed due to security restrictions. Please contact your system administrator.\n\n(Document type: %s, Operation: %s)')
+                % (self._description, operation))
+            raise exceptions.AccessError(error_message)
+
+    @api.multi
+    def _filter_access(self, operation):
+        if not self.check_access_rights(operation, raise_exception=False):
+            return self.env[self._name]
+
+        # write / unlink: valid for creator / assigned
+        valid = self.env[self._name]
         if operation in ('write', 'unlink'):
-            try:
-                self.check_access_rule(operation)
-            except exceptions.AccessError:
-                pass
-            else:
-                return
-        doc_operation = 'read' if operation == 'read' else 'write'
+            valid = self._filter_access_rule(operation, raise_exception=False)
+            if valid and valid == self:
+                return self
+
+        # compute remaining for hand-tailored rules
+        valid_sudo = valid.sudo() if valid else self.env[self._name].sudo()
+        self_sudo = self.sudo()
+        remaining = self_sudo - valid_sudo
+
+        # fall back on related document access right checks. Use the same as defined for mail.thread
+        # if available; otherwise fall back on read for read, write for other operations.
         activity_to_documents = dict()
-        for activity in self.sudo():
-            activity_to_documents.setdefault(activity.res_model, list()).append(activity.res_id)
-        for model, res_ids in activity_to_documents.items():
-            self.env[model].check_access_rights(doc_operation, raise_exception=True)
-            try:
-                self.env[model].browse(res_ids).check_access_rule(doc_operation)
-            except exceptions.AccessError:
-                raise exceptions.AccessError(
-                    _('The requested operation cannot be completed due to security restrictions. Please contact your system administrator.\n\n(Document type: %s, Operation: %s)') %
-                    (self._description, operation))
+        for activity in remaining:
+            # write / unlink: if not updating self or assigned, limit to automated activities to avoid
+            # updating other people's activities. As unlinking a document bypasses access rights checks
+            # on related activities this will not prevent people from deleting documents with activities
+            # create / read: just check rights on related document
+            if operation in ('create', 'read') or activity.automated:
+                activity_to_documents.setdefault(activity.res_model, list()).append(activity.res_id)
+        for doc_model, doc_ids in activity_to_documents.items():
+            if getattr(self.env[doc_model], '_mail_post_access', False):
```

Comment:
Well I need the value of the attribute itself (can be 'read' or 'write'). Having geattr with a default argument allow to avoid having to use hasattr + getattr.